### DISCUSSION TITLE
Predicting beyond n_forecasts  with make_future_dataframe

### DISCUSSION THREAD

**Discussion by MonacoFlori** [ourownstory/neural_prophet#1549] [Q&A - get help using NeuralProphet]
I have also a problem with the make_future_dataframe function of NeuralProphet: As soon as I add a future regressor, the function does not create a dataframe that goes into the future for more than one period (in my case a day). Therefore, I can also not add the dataframe containing the future regressors values.

I tried everything I found in the documentation and also the rest I found in the discussions but nothing works for me. If I add  n_forecasts to the model, it looses massively in accuracy, and in the end it will also not create the right dataframe.

![image](https://github.com/ourownstory/neural_prophet/assets/125356321/5f582d4e-2218-4f12-8c9a-ec361287a331)

**Comment by ourownstory** [comment]
Hi @MonacoFlori 
What you are describing here is driven by the addition of lags (AR) - not future regressor. 
Adding lagged observations (AR or lagged regressor) requires the model to have access to recent observations. 
Thus a forecast can only be made for the specified n_forecasts. 

You can make more predictions by 'unrolling' the model - predict, then add the predictions as (estimated) y values of the future to the dataframe, extend with make-future-dataframe into even further future, predict again, repeat.  You can define a loop to unroll as you like. 
We did not define this as this should be done intentionally.

And yes, the higher n_forecasts, the lower accuracy is - as you are directly regressing onto a more distant future.
I suggest you do the above suggested loop with a more reasonable n_forecasts - 0.5-1x n_lags (in your case n_forecasts=6-12) and unroll as needed. You will observe that the further you unroll, accuracy decreases.

I hope this helps!

**Reply by MonacoFlori** [reply]
Hi @ourownstory ,

thank you for your quick answer. I´m still trying to understand: If I remove the lags, is it then possible to use the make_future_dataframe with the future regressor, if I add a dataframe with (estimated) values of the future regressor for the timeframe I want to predict?
Or is the lagged regressor "included" if I add the future_regressor, because I also added values of the "future regressor" that are not in the timeframe I want to predict, but in the past, to train the model? Sorry, I´m a bit confused....

**Comment by ourownstory** [comment]
> I´m still trying to understand: If I remove the lags, is it then possible to use the make_future_dataframe with the future regressor, if I add a dataframe with (estimated) values of the future regressor for the timeframe I want to predict?

If I get your Q right, yes.


> Or is the lagged regressor "included" if I add the future_regressor, because I also added values of the "future regressor" that are not in the timeframe I want to predict, but in the past, to train the model? Sorry, I´m a bit confused....

At prediction time, the model needs any inputs that it was trained with. if if was trained with AR, then it needs the preceeding n_lags of observations before your prediction starts, and then predicts N-forecast next steps - identical to training.

Without lags, it is an entirely different model that is only dependent on timestamps (and any future regressors/events), and thus can be predicted for an arbitrary number of periods - independent of training.

**Reply by MonacoFlori** [reply]
I tried removing the lags, now I don´t get an error anymore, but I also get no prediction. I predicted the future regressor with another model, this is added to the dataframe but no prediction. I also do not understand how I can unroll the model, becuse there is no prediction at all. But how I understand it, I do need a prediction also for unrolling it. I also do not get how to add the lags.
I´m very sorry, but I just started with this so I am not too deep into the tuning of the code....

![image](https://github.com/ourownstory/neural_prophet/assets/125356321/25ad7bda-b1e6-430a-8e93-b9de3b39b682)
![image](https://github.com/ourownstory/neural_prophet/assets/125356321/55df7ce8-267c-49d9-9860-11c0a8fc4be2)

**Reply by ourownstory** [reply]
Without lags, unrolling is not necessary.
Also, setting n_forecasts and ar_layers is only relevant with lags. try removing those.

If you are getting no predictions, please share a minimal code snipped that I can use to reproduce this and see what's wrong. Thank you!

**Reply by MonacoFlori** [reply]
That is the part of the code which I think is relevant. I hope you can help me, but I do not understand how to add the future regressor. Do I also have a lagged regressor or is it not possible to use them both at the same time? I´m very sorry to bother you this much but I do not get the connection between the different params.


df = yf.download("AAPL", start="2019-01-02", end="2024-01-01")

df.drop(["Open", "High", "Low", "Adj Close", "Volume"], axis = 1, inplace=True)
df.reset_index(inplace=True, col_level=0)
df.columns = ["ds", "y"]

# Add a future regressor
df_fut = yf.download("AAPL", start="2019-01-02", end="2024-01-01")
df_fut.drop(["Open", "High", "Low", "Adj Close"], axis = 1, inplace=True)
df_fut.reset_index(inplace=True, col_level=0)
df_fut.columns = ["ds", "y", "Volume"]

m7 = NeuralProphet(n_changepoints = 12,
                   yearly_seasonality = True,
                   weekly_seasonality = True,
                   daily_seasonality = True,
                   n_lags = 20,
                  n_forecasts = 20)
set_random_seed(0)
df_train, df_test = m7.split_df(df_vol, valid_p = 0.2)
m7.add_future_regressor("Volume")
metrics_7 = m7.fit(df_train, validation_df=df_test, progress=None, freq = "D", epochs = 500)

vol_pred = yf.download("AAPL", start="2019-01-02", end="2024-01-01")
vol_pred.drop(["Open", "High", "Low", "Adj Close", "Close"], axis = 1, inplace=True)
vol_pred.reset_index(inplace=True, col_level=0)
vol_pred.columns = ["ds", "y"]

model_vol = NeuralProphet(n_changepoints = 25,
                   yearly_seasonality = True,
                   weekly_seasonality = True,
                   daily_seasonality = True,
                   n_lags = 12,
                  n_forecasts = 30)

metrics_model_vol = model_vol.fit(vol_pred)

future = model_vol.make_future_dataframe(vol_pred, n_historic_predictions = False, periods = 90)
forecast_vol = model_vol.predict(future)

def extract_yhat(forecasts, size):
  columns = forecast_vol.columns[3:]
  newframe = forecast_vol[['ds', 'yhat1']].iloc[-size:].copy()
  for col in columns:
      if 'yhat' in col:
         newframe['yhat1'] = newframe['yhat1'].fillna(forecast_vol[col])
  return newframe

volume_pred = extract_yhat(forecast_vol, 31)

volume_pred.columns = ["ds", "Volume"]

df_end = pd.concat(objs = [df_fut, volume_pred], axis = 0)

future1 = m7.make_future_dataframe(df_end, n_historic_predictions = True, periods = 30)
forecast_end = m7.predict(future1)

**Reply by MonacoFlori** [reply]
I´m sorry that the code is so much, I thought you would need it to recreate the error.
The main part of NeuralProphet would be:

m7 = NeuralProphet(n_changepoints = 12,
yearly_seasonality = True,
weekly_seasonality = True,
daily_seasonality = True,
n_lags = 20,
n_forecasts = 20)
set_random_seed(0)
m7.add_future_regressor("Volume")
metrics_7 = m7.fit(df, progress=None, freq = "D", epochs = 500)

future1 = m7.make_future_dataframe(df, regressors_df = df_vol, n_historic_predictions = True, periods = 30)
forecast = m7.predict(future1)

I am quite desperate as I need it for a project at my data science training

**Reply by ourownstory** [reply]
@MonacoFlori Very sorry, but I am currently not able to help you much with this, but you may find some of the principles I tried to point to explained in this great resource: https://otexts.com/fpp3/
